### PR TITLE
CSHARP-4886: Mark API that will be removed in 3.0 as obsolete: MongoDB.Bson

### PR DESCRIPTION
**Pull request by BorisDog**
Additional entities to consider: 
Utf8Helper
IBsonChunk
IBsonChunkSource
IByteBuffer
IChildSerializerConfigurable

### CODE REVIEW

File: src/MongoDB.Bson/Serialization/AttributedSerializationProvider.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides serializers based on an attribute.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/AttributedSerializationProvider.cs:26]: Really?

File: src/MongoDB.Bson/Serialization/BsonObjectModelSerializationProvider.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides serializers for BsonValue and its derivations.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/BsonObjectModelSerializationProvider.cs:26]: Really?

File: src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs [hunk 3]
```diff
@@ -28,6 +28,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides serializers for collections.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs:31]: Really?

File: src/MongoDB.Bson/Serialization/Serializers/TupleSerializers.cs [hunk 4]
```diff
@@ -26,7 +26,7 @@ namespace MongoDB.Bson.Serialization.Serializers
     public interface IBsonTupleSerializer
     {
         /// <summary>
-        /// Gets ths serializer for an item.
+        /// Gets this serializer for an item.
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Serializers/TupleSerializers.cs:29]: /// Gets the serializer for an item.

File: src/MongoDB.Bson/Serialization/DiscriminatedInterfaceSerializationProvider.cs [hunk 5]
```diff
@@ -22,6 +22,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides a serializer for interfaces.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/DiscriminatedInterfaceSerializationProvider.cs:25]: Really?

File: src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs [hunk 6]
```diff
@@ -24,6 +24,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// A helper class used to create and compile delegates for creator maps.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs:27]: Really?

File: src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs [hunk 7]
```diff
@@ -78,13 +78,22 @@ public static BsonBinaryDataGuidGenerator StandardInstance
         }
 
         /// <summary>
-        /// Gets an instance of BsonBinaryDataGuidGenerator for Unspecifed GuidRepresentation.
+        /// Gets an instance of BsonBinaryDataGuidGenerator for Unspecified GuidRepresentation.
         /// </summary>
+        [Obsolete("This method will be removed in later release. Please use UnspecifiedInstance instead")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs:83]: Really?

File: src/MongoDB.Bson/Serialization/IdGenerators/ZeroIdChecker.cs [hunk 8]
```diff
@@ -22,6 +22,7 @@ namespace MongoDB.Bson.Serialization.IdGenerators
     /// </summary>
     /// <typeparam name="T">The type of the Id.</typeparam>
     // TODO: is it worth trying to remove the dependency on IEquatable<T>?
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/IdGenerators/ZeroIdChecker.cs:25]: Really?

File: src/MongoDB.Bson/Serialization/PrimitiveSerializationProvider.cs [hunk 9]
```diff
@@ -26,6 +26,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides serializers for primitive types.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/PrimitiveSerializationProvider.cs:29]: Really?

File: src/MongoDB.Bson/Serialization/TypeMappingSerializationProvider.cs [hunk 10]
```diff
@@ -22,6 +22,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Represents a serialization provider based on a mapping from value types to serializer types.
     /// </summary>
+    [Obsolete("This class will be removed in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/TypeMappingSerializationProvider.cs:25]: Really?

File: tests/BuildProps/Tests.Build.props [hunk 11]
```diff
@@ -54,4 +54,25 @@
     </PackageReference>
   </ItemGroup>
 
+  <PropertyGroup>
+    <NoWarn>
+      1701;1702; <!--https://github.com/dotnet/roslyn/issues/19640-->
```
[the remark this answers is not in the document]
Comment (rstam) [reply tests/BuildProps/Tests.Build.props:59]: This is a sign that these changes will be painful for our users.

File: src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj [hunk 12]
```diff
@@ -17,7 +17,10 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <NoWarn>NU5100</NoWarn>
+    <NoWarn>
+      NU5100;
+      CS0618 <!--Temporary disable obsolete warnings due the massive volume of warnings-->
+    </NoWarn>
```
Comment (rstam) [inline src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj:23]: This is a sign that these changes will be painful for our users.

**Review by rstam — CHANGES_REQUESTED** [review]
First: I think we are doing this way too prematurely. Am waiting for @JamesKovacs to get back to revisit this question with him.

Second: I think there is some uncertainty about *what* exactly is going to be deprecated. We won't really know for sure until we are close to being done with 3.0 and see what ended up changing and what didn't.

File: src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs (hunk 7 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs:83]: Yes, directing to use `UnspecifiedInstance` instead.

File: src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs [hunk 13]
```diff
@@ -78,13 +78,22 @@ public static BsonBinaryDataGuidGenerator StandardInstance
...
         public static BsonBinaryDataGuidGenerator UnspecifedInstance
         {
             get { return __unspecifiedInstance; }
         }
 
+        /// <summary>
+        /// Gets an instance of BsonBinaryDataGuidGenerator for Unspecified GuidRepresentation.
+        /// </summary>
+        public static BsonBinaryDataGuidGenerator UnspecifiedInstance
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/IdGenerators/BsonBinaryDataGuidGenerator.cs:92]: Yes, fixing a typo.

File: src/MongoDB.Bson/Serialization/IdGenerators/ZeroIdChecker.cs (hunk 8 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/IdGenerators/ZeroIdChecker.cs]: Reverted.

File: src/MongoDB.Bson/Serialization/PrimitiveSerializationProvider.cs (hunk 9 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/PrimitiveSerializationProvider.cs:29]: Will our user be needing our built-in serializer providers?

File: src/MongoDB.Bson/Serialization/Serializers/TupleSerializers.cs (hunk 4 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/Serializers/TupleSerializers.cs]: Thanks, done.

File: src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj (hunk 12 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj:23]: But hopefully significantly less painful than just removing API in 3.0 without prior warning.

File: tests/BuildProps/Tests.Build.props (hunk 11 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/BuildProps/Tests.Build.props]: But hopefully significantly less painful than just removing API in 3.0 without prior warning.

**Review by BorisDog — COMMENTED** [review]
Addressed some of the comments.
Please advise what else needs to be deprecated.

File: src/Directory.Build.props [hunk 14]
```diff
@@ -51,4 +51,10 @@
     <None Include="..\..\LICENSE.md" Pack="true" PackagePath="$(PackageLicenseFile)" />
     <None Include="..\..\packageIcon.png" Pack="true" PackagePath="" />
   </ItemGroup>
+
+  <PropertyGroup>
+    <NoWarn>
+      CS0618 <!--Temporary disable obsolete warnings due the massive volume of warnings-->
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/Directory.Build.props:57]: I think we should have a goal that users who stay on 2.x don't have to deal with these warnings at all.

File: src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs (hunk 3 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs:31]: Users can write custom serialization providers and in general configure any mix of serialization providers in any order. To do so the providers need to be public.

Removing these classes (or making them internal) removes a feature.

File: src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs (hunk 6 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs:27]: I thought when you said "removed" you meant "removed".

It probably can be made `internal`.

File: src/MongoDB.Bson/ObjectModel/BsonSymbol.cs [hunk 15]
```diff
@@ -1,182 +1,183 @@
...
-        /// Returns a string representation of the value.
-        /// </summary>
-        /// <returns>A string representation of the value.</returns>
-        public override string ToString()
-        {
-            return _name;
-        }
-    }
-}
+﻿/* Copyright 2010-present MongoDB Inc.
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonSymbol.cs:1]: What happened here?

Why does the entire file appear to be modified?

Same for several other files below.

File: src/MongoDB.Bson/Serialization/Serializers/BsonValueSerializer.cs [hunk 16]
```diff
@@ -1,155 +1,155 @@
...
-                case BsonType.String: BsonStringSerializer.Instance.Serialize(context, (BsonString)value); break;
-                case BsonType.Symbol: BsonSymbolSerializer.Instance.Serialize(context, (BsonSymbol)value); break;
-                case BsonType.Timestamp: BsonTimestampSerializer.Instance.Serialize(context, (BsonTimestamp)value); break;
-                case BsonType.Undefined: BsonUndefinedSerializer.Instance.Serialize(context, (BsonUndefined)value); break;
-                default: throw new BsonInternalException("Invalid BsonType.");
-            }
-        }
-    }
-}
+/* Copyright 2010-present MongoDB Inc.
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Serializers/BsonValueSerializer.cs:1]: Hard to tell if you changed anything in this file or not...

Same for several other files.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs (hunk 3 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs]: Do we really need the predefined providers to be public, is our intent to fully support that API going forward?
Looks like they all are added in `CreateSerializerRegistry` and user has no control over it.

File: src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs (hunk 6 above)
Comment (BorisDog) [reply src/MongoDB.Bson/Serialization/CreatorMapDelegateCompiler.cs]: Updated the comments.

File: src/MongoDB.Bson/Serialization/AttributedSerializationProvider.cs [hunk 17]
```diff
@@ -23,6 +23,7 @@ namespace MongoDB.Bson.Serialization
     /// <summary>
     /// Provides serializers based on an attribute.
     /// </summary>
+    [Obsolete("This class will be removed from public API in later release.")]
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/AttributedSerializationProvider.cs:26]: Why are we removing serialization providers, including this one?

The user is supposed to be able to configure the serialization providers they want to use.

Comment (rstam) [reply src/MongoDB.Bson/Serialization/AttributedSerializationProvider.cs:26]: Same for similar obsoletions below...

File: src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj (hunk 12 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj:23]: We can't leave this in here.

This is going to hide important warnings even from ourselves!!!

File: src/Directory.Build.props (hunk 14 above)
[the remark this answers is not in the document]
Comment (rstam) [reply src/Directory.Build.props:57]: We can't leave this in here.

This is going to hide important warnings even from ourselves!!!

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by BorisDog** [comment]
Stale PR.

**Pull request closed without merging**